design/ceph/admission-controller.md

```diff
@@ -0,0 +1,93 @@
+* Modifying the webhook.yaml to inject CA bundle onto the ValidatingWebhookConfig
+* Creating a webhook deployment, service, config on the cluster and deploying the entire webhook setup
+* Now our webhook server has been configured, deployed and ready to intercept requests based on the webhookconfig.
+* **All the above resources will be created in rook-ceph namespace hence it should be created in prior**
+
+Using the above approach, the dev/admins will have the responsibility of rotating the certificates when they expire.
+
+Below is an excerpt of what a ValidatingWebhookConfig looks like 
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: demo-webhook
+webhooks:
+  - name: webhook-server.webhook-demo.svc
+    clientConfig:
+      service:
+        name: webhook-server
+        namespace: rook-ceph
+        path: "/validate"
+      caBundle: ${CA_PEM_B64}
+    rules:
+      - operations: [ "CREATE" ]
+        apiGroups: [""]
+        apiVersions: ["v1"]
+        resources: ["pods"]
+```
+
+We can make changes to the above values according to intercept based on whether a resource is being updated/deleted/created or change the type of resource or the request path which will be sent to the server.
+####Production
+The onus of providing proper certificates and CA bundle for the server would be on the user. The following example would provide how it would work in an OCP cluster environment
+#####Openshift Container Platform
+![AC Image](./rook ac-openshift.png)
+
+In the OCP cluster, The management of certificates would be completely done by the Openshift Service CA Operator. 
+
+On annotating any service with 
+`service.beta.openshift.io/serving-cert-secret-name:<secret-name>` 
+
+The operator automatically generates a signed serving certificate and key pair(tls.crt and tls.key) into a secret in the same namespace as the service. 
+This secret will be mounted onto the webhook pod. Our server would be looking for the files (tls.key and tls.crt) and automatically set them for TLS communication which is why we had hardcoded earlier for dev and testing env.
+For injecting a CA bundle onto the ValidatingWebhookConfig, we can annotate it with
+`service.beta.openshift.io/inject-cabundle=true`
+
+The operator on detecting this will set the ca bundle field in our config with a base64url-encoded CA signing bundle. These certificates will be automatically rotated and renewed by the service CA operator. This can also be done manually through the commands which can be found in the operator manual.
+
+**All the annotated components will be provided in a separate file called openshift-webhook.yaml**
+
+###Starting the Webhook Server
+
+The webhook server binaries will be bundled with rook operator image. An environment variable will be set in rook operator
+
+```ROOK_ADMISSION_CONTROLLER : TRUE```
```

Comment:
Instead of adding a new setting like this, what if the operator automatically started the admission controller if it finds the expected secret with the tls key?

Reply:
That is good idea. I would like to have a few more discussions on this today. I'll schedule a call.

Reply:
Made the changes for this in the design doc. Please review